gateway-js/src/index.ts

```diff
@@ -425,135 +416,143 @@ export class ApolloGateway implements GraphQLService {
+      this.updateWithSupergraphSdl({
+        supergraphSdl,
+        id: this.getIdForSupergraphSdl(supergraphSdl),
+      });
+    } finally {
+      // if update fails, we still want to go back to `loaded` state
+      this.state = { phase: 'loaded' };
     }
+  }
 
-    await this.maybePerformServiceHealthCheck(result);
-
-    this.compositionMetadata = result.compositionMetadata;
-    this.serviceDefinitions = result.serviceDefinitions;
-
-    const { schema, supergraphSdl } = this.createSchemaFromServiceList(
-      result.serviceDefinitions,
-    );
+  private async externalSubgraphHealthCheckCallback(supergraphSdl: string) {
+    const parsedSupergraphSdl =
+      supergraphSdl === this.supergraphSdl
+        ? this.parsedSupergraphSdl
+        : parse(supergraphSdl);
+
+    const serviceList = this.serviceListFromSupergraphSdl(parsedSupergraphSdl!);
+    // Here we need to construct new datasources based on the new schema info
+    // so we can check the health of the services we're _updating to_.
+    const serviceMap = serviceList.reduce((serviceMap, serviceDef) => {
+      serviceMap[serviceDef.name] = {
+        url: serviceDef.url,
+        dataSource: this.createDataSource(serviceDef),
+      };
+      return serviceMap;
+    }, Object.create(null) as DataSourceMap);
 
-    if (!supergraphSdl) {
-      this.logger.error(
-        "A valid schema couldn't be composed. Falling back to previous schema.",
+    try {
+      await this.serviceHealthCheck(serviceMap);
+    } catch (e) {
+      throw new Error(
+        'The gateway subgraphs health check failed. Updating to the provided ' +
+          '`supergraphSdl` will likely result in future request failures to ' +
+          'subgraphs. The following error occurred during the health check:\n' +
+          e.message,
       );
-    } else {
-      this.updateWithSchemaAndNotify(schema, supergraphSdl);
-
-      if (this.experimental_didUpdateComposition) {
-        this.experimental_didUpdateComposition(
-          {
-            serviceDefinitions: result.serviceDefinitions,
-            schema,
-            ...(this.compositionMetadata && {
-              compositionMetadata: this.compositionMetadata,
-            }),
-          },
-          previousServiceDefinitions &&
-            previousSchema && {
-              serviceDefinitions: previousServiceDefinitions,
-              schema: previousSchema,
-              ...(previousCompositionMetadata && {
-                compositionMetadata: previousCompositionMetadata,
-              }),
-            },
-        );
-      }
     }
   }
 
-  private async updateWithSupergraphSdl(
-    result: SupergraphSdlUpdate,
-  ): Promise<void> {
-    if (result.id === this.compositionId) {
-      this.logger.debug('No change in composition since last check.');
-      return;
-    }
+  private externalGetDataSourceCallback({
+    name,
+    url,
+  }: ServiceEndpointDefinition) {
+    return this.createAndCacheDataSource({ name, url });
```

Comment:
Not introduced by this PR but `getOrCreateDataSource` is a more appropriate name

Reply:
437edd6f